### PR TITLE
Add TMA tiling utils into creed

### LINKED ISSUE #12: Add tiling scripts to the repo
Add the tiling notebooks currently in `ark` along with `tiling_utils.py` over.

### PR DESCRIPTION
**Pull request by alex-l-kong**
Closes #12. Now that the tiling processes have been finalized, we add them to `creed-helper` to ensure they can be accessed on the MibiScope computer.

We'll start with the TMA tiling utils first, then add the tiled regions/non-TMA in afterward.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
Copied over the `tiling_utils.py` and `example_tma_fov_generate.ipynb` over. I'll be working on setting up `Docker` and `Travis` separately; `base_dir` in the notebook is just a dummy value for the time being.

**Comment by ngreenwald** [comment]
travis/pytest/coveralls yes, docker no, since this will be on windows machines

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:275]: same thing here

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looking good. A few more naming suggestions

File: creed/tiling_utils.py
Comment (alex-l-kong) [reply creed/tiling_utils.py:126]: I think `region_corners` sounds good.

File: creed/tiling_utils.py
Comment (alex-l-kong) [reply creed/tiling_utils.py:157]: Are you talking about the `frameSizePixels` key? If that's the case, a couple follow-ups:

1. Does the commercial instrument take `frameSizeMicrons` as well or does it need to be `frameSizePixels`?
2. More to the above point, if `frameSizePixels` is the only supported spec, using the fiducial constants in reverse to convert pixels to microns should work.
3. The above points are for tiled regions (non-TMA); how will this affect TMA tiling, if at all? Can we ignore `frameSizePixels` and assume the user will just want to specify a specific number of FOVs along each coordinate axis, or should we determine programmatically how many FOVs along the x- and y-axes we should use based on `frameSizePixels` (or `frameSizeMicrons` if that can be specified)?

File: creed/tiling_utils.py
Comment (alex-l-kong) [reply creed/tiling_utils.py:263]: How does something like this look in comparison:

```
`moly_interval = read_tiling_param(
    "Enter the FOV interval size to insert Moly points. If you do not want to insert Moly points at a FOV interval, enter 0.",
    "Error: moly interval must be 0 or a positive integer",
    lambda mi: mi >= 0,
    dtype=int
)

if moly_interval > 0:
    tiling_params['moly_interval'] = moly_interval
```

File: creed/tiling_utils.py
Comment (alex-l-kong) [reply creed/tiling_utils.py:273]: We had agreed to do it this way when the tiling notebook was originally spun up, but I can change it back to loading it in the function.

**Comment by ngreenwald** [comment]
I changed my mind, we should add back the saving of the automatic TMA coords to make it easier to reproduce the output at a later date

File: creed/tiling_utils.py
Comment (ngreenwald) [reply creed/tiling_utils.py:157]: Lets discuss at standup

File: creed/tiling_utils.py
Comment (ngreenwald) [reply creed/tiling_utils.py:251]: looks good

File: creed/tiling_utils.py
Comment (ngreenwald) [reply creed/tiling_utils.py]: Which one do you think makes more sense?

Comment (alex-l-kong) [reply creed/tiling_utils.py]: To me, it makes more sense to load it in `generate_tiled_region_fov_list`. It's a little awkward as it is right now and this ensures we only load in the Moly point when we actually need it. It also matches up with what we do in `remap_and_reorder_fovs`.

**Comment by alex-l-kong** [comment]
> I changed my mind, we should add back the saving of the automatic TMA coords to make it easier to reproduce the output at a later date

To clarify, you mean creating a full-fledged JSON run file containing each automatic TMA coord like we did earlier? Or would simply saving the mapping from automatic FOV name be coordinate enough?

**Comment by ngreenwald** [comment]
Saving the current pared down automatic JSON that just contains names and centroid coordinates. 

Basically, we want someone in 3 months to be able to recreate the same interface. So they'll need to load their manual names, the automatic names we created, and the final mapping they decided on, so they can inspect the results

File: creed/tiling_utils.py
Comment (alex-l-kong) [reply creed/tiling_utils.py:157]: Per our standup today, we'll open a separate issue to handle this after the TMA process is merged in. We'll be using `fovSizeMicrons` to space the tiled regions (non-TMA) version.

File: creed/tiling_utils.py
Comment (ngreenwald) [inline creed/tiling_utils.py:231]:
```suggestion
        "Insert a moly point between each tiled region? Y/N: ",
```

File: creed/tiling_utils.py
Comment (ngreenwald) [inline creed/tiling_utils.py:243]: Change to: "Insert moly points between FOVs? If yes, enter the number of FOVs between each Moly point. If no, enter 0. "

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, minor comments only

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, re-request once the rhombus branch is merged in

**Comment by alex-l-kong** [comment]
After we merge this into `master`, I'll update the `tma_validate` PR to merge directly into `main`.

File: creed/tiling_utils.py
Comment (ngreenwald) [inline creed/tiling_utils.py:15]: merge main into this branch update creed references to toffy

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:154]: These assert statements can be condensed into a for loop over the tuple of parameter (i.e. region_start_x), base_value (i.e. 100)

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:195]: add a `parameterize = pytest.mark.paramtrize` at the top to shorten calls

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:106]: This function should be moved from ark to this repo.

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:124]: This should be a for loop. In addition, the values that are being evaluated should be passed from the function that creates the data, rather than being hardcoded here

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:179]: Same thing here

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:231]: Same thing here.

File: creed/tiling_utils_test.py
Comment (ngreenwald) [reply creed/tiling_utils_test.py:195]: Given that this global is only used for this function, it makes more sense to just use [0, 1] here instead

File: creed/tiling_utils_test_cases.py
Comment (ngreenwald) [inline creed/tiling_utils_test_cases.py:116]: This comment is useful, but it would be better to define the x_increment using the relationship you describe, so that if coords changes this value changes automatically too

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:413]: You can mark the cases that are expected to fail with `xfail`; see the rosetta PR. This will allow you to combine these functions together

File: creed/tiling_utils_test.py
Comment (ngreenwald) [inline creed/tiling_utils_test.py:436]: This should be in the cases documentation, rather than here

File: creed/tiling_utils_test_cases.py
Comment (ngreenwald) [inline creed/tiling_utils_test_cases.py:10]: Putting something in the cases file is useful to hide complicated functions. For simple lists of values, having them in the decorator is better, because than its easy to see what's being tested

File: creed/tiling_utils_test_cases.py
Comment (ngreenwald) [inline creed/tiling_utils_test_cases.py:21]: this is a good example one that does belong in here

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Make sure you merge main into this branch to get the toffy changes. 

Most of my comments are about the testing. Feel free to address those in the validation PR if you'd prefer to merge this in, or do it in here if it'll be easier to keep track of

File: creed/tiling_utils_test.py
Comment (alex-l-kong) [reply creed/tiling_utils_test.py]: Were you thinking of something more or less like:

```
# define the starting values to increment from for each param
base_vals = {
    'region_start_x': 100,
    'region_start_y': 200,
    'fov_num_x': 3,
    'fov_num_y': 6,
    'x_fov_size': 5,
    'y_fov_size': 10
}

# the keys are going to be the same for all regions of sample_region_params
for i in list(zip(sample_region_params[0].items()))
    for param, val in list(sample_region_params[i].items()):
        assert val == base_vals[param] * (i + 1)
```

Comment (ngreenwald) [reply creed/tiling_utils_test.py]: Yup, and you could use base_vals to generate sample_region_inputs as well

**Comment by alex-l-kong** [comment]
Ok, refactoring's done. Had a couple TODO questions but everything else has been addressed.

I pulled out some tests from `parametrize_with_cases` to `parametrize` with a specific `marks=[xfail]` argument where it was easier to test that way.

**Review by ngreenwald — APPROVED** [review]
Sweet, looks good

**Pull request merged**
